src/compute/substring.rs

```diff
@@ -54,41 +41,30 @@ fn utf8_substring<O: Offset>(array: &Utf8Array<O>, start: O, length: &Option<O>)
         };
 
         let mut iter_chars = str_val.char_indices();
-        if let Some((start_idx, _char)) = iter_chars.nth(start) {
-            // length till end of str
+        if let Some((start_idx, _)) = iter_chars.nth(start) {
+            // length of the str
             let len_end = str_val.len() - start_idx;
 
             // length to slice
-            let length = length.map(|v| v.to_usize()).unwrap_or(len_end);
+            let length = length.unwrap_or(len_end);
 
-            // index of the char with offset `start`, and length: `length`
+            if length == 0 {
+                return "";
+            }
+            // compute
             let end_idx = iter_chars
                 .nth(length.saturating_sub(1))
                 .map(|(idx, _)| idx)
                 .unwrap_or(str_val.len());
-            if length != 0 {
-                debug_assert!(std::str::from_utf8(
-                    &values[windows[0].to_usize() + start_idx..windows[0].to_usize() + end_idx]
-                )
-                .is_ok());
-                new_values.extend_from_slice(
-                    &values[windows[0].to_usize() + start_idx..windows[0].to_usize() + end_idx],
-                );
-            }
+
+            &str_val[start_idx..end_idx]
+        } else {
+            ""
         }
-        new_offsets.push(O::from_usize(new_values.len()).unwrap());
     });
 
-    // Safety:
-    // we deal with valid utf8
-    unsafe {
-        Utf8Array::<O>::from_data_unchecked(
-            array.data_type().clone(),
-            new_offsets.into(),
-            new_values.into(),
-            validity.cloned(),
-        )
-    }
+    let new = Utf8Array::<O>::from_trusted_len_values_iter(iter);
```

Comment:
And because the iterator has `Item=&str`, we don't have to do utf8 checks right?

Reply:
exactly. :)